diff --git a/Makefile b/Makefile
index 39a99d7..63849cf 100644
--- a/Makefile
+++ b/Makefile
@@ -120,6 +120,7 @@ UPROGS=\
 	$U/_echo\
 	$U/_forktest\
 	$U/_grep\
+	$U/_history\
 	$U/_init\
 	$U/_kill\
 	$U/_ln\
@@ -128,6 +129,7 @@ UPROGS=\
 	$U/_rm\
 	$U/_sh\
 	$U/_stressfs\
+	$U/_trace\
 	$U/_usertests\
 	$U/_grind\
 	$U/_wc\
diff --git a/README b/README
index ed8bba5..bae9fde 100644
--- a/README
+++ b/README
@@ -1,49 +1,92 @@
-xv6 is a re-implementation of Dennis Ritchie's and Ken Thompson's Unix
-Version 6 (v6).  xv6 loosely follows the structure and style of v6,
-but is implemented for a modern RISC-V multiprocessor using ANSI C.
-
-ACKNOWLEDGMENTS
-
-xv6 is inspired by John Lions's Commentary on UNIX 6th Edition (Peer
-to Peer Communications; ISBN: 1-57398-013-7; 1st edition (June 14,
-2000)).  See also https://pdos.csail.mit.edu/6.1810/, which provides
-pointers to on-line resources for v6.
-
-The following people have made contributions: Russ Cox (context switching,
-locking), Cliff Frey (MP), Xiao Yu (MP), Nickolai Zeldovich, and Austin
-Clements.
-
-We are also grateful for the bug reports and patches contributed by
-Takahiro Aoyagi, Silas Boyd-Wickizer, Anton Burtsev, carlclone, Ian
-Chen, Dan Cross, Cody Cutler, Mike CAT, Tej Chajed, Asami Doi,
-eyalz800, Nelson Elhage, Saar Ettinger, Alice Ferrazzi, Nathaniel
-Filardo, flespark, Peter Froehlich, Yakir Goaron, Shivam Handa, Matt
-Harvey, Bryan Henry, jaichenhengjie, Jim Huang, Matúš Jókay, John
-Jolly, Alexander Kapshuk, Anders Kaseorg, kehao95, Wolfgang Keller,
-Jungwoo Kim, Jonathan Kimmitt, Eddie Kohler, Vadim Kolontsov, Austin
-Liew, l0stman, Pavan Maddamsetti, Imbar Marinescu, Yandong Mao, Matan
-Shabtay, Hitoshi Mitake, Carmi Merimovich, Mark Morrissey, mtasm, Joel
-Nider, Hayato Ohhashi, OptimisticSide, Harry Porter, Greg Price, Jude
-Rich, segfault, Ayan Shafqat, Eldar Sehayek, Yongming Shen, Fumiya
-Shigemitsu, Cam Tenny, tyfkda, Warren Toomey, Stephen Tu, Rafael Ubal,
-Amane Uehara, Pablo Ventura, Xi Wang, WaheedHafez, Keiichi Watanabe,
-Nicolas Wolovick, wxdao, Grant Wu, Jindong Zhang, Icenowy Zheng,
-ZhUyU1997, and Zou Chang Wei.
-
-
-The code in the files that constitute xv6 is
-Copyright 2006-2022 Frans Kaashoek, Robert Morris, and Russ Cox.
-
-ERROR REPORTS
-
-Please send errors and suggestions to Frans Kaashoek and Robert Morris
-(kaashoek,rtm@mit.edu).  The main purpose of xv6 is as a teaching
-operating system for MIT's 6.1810, so we are more interested in
-simplifications and clarifications than new features.
-
-BUILDING AND RUNNING XV6
-
-You will need a RISC-V "newlib" tool chain from
-https://github.com/riscv/riscv-gnu-toolchain, and qemu compiled for
-riscv64-softmmu.  Once they are installed, and in your shell
-search path, you can run "make qemu".
+# Working with xv6
+
+## Creating a new system call program
+
+1. Create a new file <new_system_call_name> in `user/` directory.
+
+    The new file should contain the following:
+    ```c
+        #include "kernel/param.h"
+        #include "kernel/types.h"
+        #include "kernel/stat.h"
+        #include "user/user.h"
+
+        int main(int argc, char *argv[]) {
+            // code
+            exit(exit_status);
+        }
+    ```
+    **For example, see the `user/trace.c` file**
+
+    `Note:` The file name should be as same as the command name that you want to use to execute the system call.
+
+4. Add the new system call command name to `user/Makefile` as follows:
+    ```makefile
+        UPROGS = \
+            ...
+            $U/_<new_command_name>\
+            ...
+    ```
+    **eg: `$U/_trace \`**
+
+    This will allow the `make qemu` to enable the system call.
+
+    `Note:` The <new_command_name> part should be the same as the file name created in step 1.
+
+3. Add an entry to `user/usys.pl`.
+
+    ```c
+        entry("<new_system_call_name>");
+    ```
+
+    **eg: `entry("trace");`**
+
+    This will route the system call into the kernel.
+
+    `Note:` The <new_system_call_name> part does not have to be the same as the file name created in step 1. But all the names stated in the following steps should be the same as this one.
+
+4. Add the prototype of the new system call in `user/user.h`.
+
+    **eg: `int trace(int);`**
+
+5. Define a system call number in `kernel/syscall.h` as follows:
+    ```c
+        #define SYS_<new_system_call_name> <new_system_call_number>
+    ```
+    **eg: `#define SYS_trace 22`**
+
+6. Define the system call function in `kernel/proc.c`
+    ```c
+        <return type>
+        <new_system_call_name>(<args>) {
+            // code
+        }
+    ```
+    **eg: see the `trace()` function in `kernel/proc.c`**
+
+7. Open `kernel/sysproc.c` and add the handler function that will execute the system call which has just been defined in step 6.
+    ```c
+        <return type>
+        sys_<new_system_call_name>(void) {
+            // code
+        }
+    ```
+    **eg: see the `sys_trace()` function in `kernel/sysproc.c`**
+8. In `kernel/syscall.c`,
+    - extern the new system call handler function.
+
+        ```c
+            extern int sys_<new_system_call_name>(void);
+        ```
+
+        **eg: `extern int sys_trace(void);`**
+
+    - Add the new system call name to the `syscalls` array.
+
+        ```c
+            [SYS_<new_system_call_name>] sys_<new_system_call_name>,
+        ```
+
+        **eg: `[SYS_trace] sys_trace,`**
+
+9. Make required changes in other functions and files as required.
\ No newline at end of file
diff --git a/compile_flags.txt b/compile_flags.txt
new file mode 100644
index 0000000..60a4aad
--- /dev/null
+++ b/compile_flags.txt
@@ -0,0 +1 @@
+-I.
diff --git a/kernel/defs.h b/kernel/defs.h
index a3c962b..995936c 100644
--- a/kernel/defs.h
+++ b/kernel/defs.h
@@ -64,6 +64,7 @@ void*           kalloc(void);
 void            kfree(void *);
 void            kinit(void);
 
+
 // log.c
 void            initlog(int, struct superblock*);
 void            log_write(struct buf*);
@@ -106,6 +107,8 @@ void            yield(void);
 int             either_copyout(int user_dst, uint64 dst, void *src, uint64 len);
 int             either_copyin(void *dst, int user_src, uint64 src, uint64 len);
 void            procdump(void);
+int             get_trace(int);
+
 
 // swtch.S
 void            swtch(struct context*, struct context*);
@@ -140,6 +143,9 @@ void            argaddr(int, uint64 *);
 int             fetchstr(uint64, char*, int);
 int             fetchaddr(uint64, uint64*);
 void            syscall();
+void            init_stats_lock();
+void             update_history(int, int, const char*);
+int             get_history(int, uint64 addr);
 
 // trap.c
 extern uint     ticks;
diff --git a/kernel/main.c b/kernel/main.c
index f0d3171..67daf40 100644
--- a/kernel/main.c
+++ b/kernel/main.c
@@ -4,6 +4,8 @@
 #include "riscv.h"
 #include "defs.h"
 
+#include "spinlock.h"
+
 volatile static int started = 0;
 
 // start() jumps here in supervisor mode on all CPUs.
@@ -30,6 +32,7 @@ main()
     virtio_disk_init(); // emulated hard disk
     userinit();      // first user process
     __sync_synchronize();
+    init_stats_lock();
     started = 1;
   } else {
     while(started == 0)
diff --git a/kernel/proc.c b/kernel/proc.c
index 959b778..bfc628f 100644
--- a/kernel/proc.c
+++ b/kernel/proc.c
@@ -124,6 +124,7 @@ allocproc(void)
 found:
   p->pid = allocpid();
   p->state = USED;
+  p->traced_pid = 0;
 
   // Allocate a trapframe page.
   if((p->trapframe = (struct trapframe *)kalloc()) == 0){
@@ -146,6 +147,9 @@ found:
   p->context.ra = (uint64)forkret;
   p->context.sp = p->kstack + PGSIZE;
 
+  // Initialize traced_pid to 0
+  // p->traced_pid = 0;
+
   return p;
 }
 
@@ -681,3 +685,11 @@ procdump(void)
     printf("\n");
   }
 }
+
+// Traces the proces
+int
+get_trace(int pid) {
+  struct proc *p = myproc();
+  p->traced_pid = pid;
+  return 0;
+}
diff --git a/kernel/proc.h b/kernel/proc.h
index d021857..3faa9df 100644
--- a/kernel/proc.h
+++ b/kernel/proc.h
@@ -104,4 +104,5 @@ struct proc {
   struct file *ofile[NOFILE];  // Open files
   struct inode *cwd;           // Current directory
   char name[16];               // Process name (debugging)
+  int traced_pid;              // Process ID of the process being traced
 };
diff --git a/kernel/syscall.c b/kernel/syscall.c
index ed65409..b67823c 100644
--- a/kernel/syscall.c
+++ b/kernel/syscall.c
@@ -6,6 +6,26 @@
 #include "proc.h"
 #include "syscall.h"
 #include "defs.h"
+#include "stdbool.h"
+
+
+#define ARG_INT  0
+#define ARG_PTR  1
+#define ARG_STR  2
+
+struct syscall_args {
+  int syscall_num;
+  union {
+    int int_arg;
+    void* ptr_arg;
+    char str_arg[16];
+  } args[6];
+  int type[6];
+  bool exists[6];
+};
+
+struct syscall_args arg_array[24];
+
 
 // Fetch the uint64 at addr from the current process.
 int
@@ -53,30 +73,45 @@ argraw(int n)
 }
 
 // Fetch the nth 32-bit system call argument.
-void
-argint(int n, int *ip)
-{
+void argint(int n, int *ip) {
+  struct proc *p = myproc();
+  *ip = argraw(n);
+  arg_array[p->pid].args[n].int_arg = *ip;
+  arg_array[p->pid].type[n] = ARG_INT;
+  arg_array[p->pid].exists[n] = true;
+}
+
+void argaddr(int n, uint64 *ip) {
+  struct proc *p = myproc();
   *ip = argraw(n);
+  arg_array[p->pid].args[n].ptr_arg = (void*)*ip;
+  arg_array[p->pid].type[n] = ARG_PTR;
+  arg_array[p->pid].exists[n] = true;
 }
 
+
 // Retrieve an argument as a pointer.
 // Doesn't check for legality, since
 // copyin/copyout will do that.
-void
-argaddr(int n, uint64 *ip)
-{
-  *ip = argraw(n);
-}
+
 
 // Fetch the nth word-sized system call argument as a null-terminated string.
 // Copies into buf, at most max.
 // Returns string length if OK (including nul), -1 if error.
-int
-argstr(int n, char *buf, int max)
-{
+int argstr(int n, char *buf, int max) {
   uint64 addr;
+  struct proc *p = myproc();
   argaddr(n, &addr);
-  return fetchstr(addr, buf, max);
+
+  int length = fetchstr(addr, buf, max);
+  if (length >= 0) {
+    strncpy(arg_array[p->pid].args[n].str_arg, buf, length);
+    arg_array[p->pid].args[n].str_arg[length] = '\0'; // Null-terminate the string
+    arg_array[p->pid].type[n] = ARG_STR;
+    arg_array[p->pid].exists[n] = true;
+  }
+
+  return length;
 }
 
 // Prototypes for the functions that handle system calls.
@@ -101,6 +136,8 @@ extern uint64 sys_unlink(void);
 extern uint64 sys_link(void);
 extern uint64 sys_mkdir(void);
 extern uint64 sys_close(void);
+extern uint64 sys_trace(void);
+extern uint64 sys_history(void);
 
 // An array mapping syscall numbers from syscall.h
 // to the function that handles the system call.
@@ -126,19 +163,131 @@ static uint64 (*syscalls[])(void) = {
 [SYS_link]    sys_link,
 [SYS_mkdir]   sys_mkdir,
 [SYS_close]   sys_close,
+[SYS_trace]   sys_trace,
+[SYS_history] sys_history,
+};
+
+char* syscall_names[] = {
+[SYS_fork]    "fork",
+[SYS_exit]    "exit",
+[SYS_wait]    "wait",
+[SYS_pipe]    "pipe",
+[SYS_read]    "read",
+[SYS_kill]    "kill",
+[SYS_exec]    "exec",
+[SYS_fstat]   "fstat",
+[SYS_chdir]   "chdir",
+[SYS_dup]     "dup",
+[SYS_getpid]  "getpid",
+[SYS_sbrk]    "sbrk",
+[SYS_sleep]   "sleep",
+[SYS_uptime]  "uptime",
+[SYS_open]    "open",
+[SYS_write]   "write",
+[SYS_mknod]   "mknod",
+[SYS_unlink]  "unlink",
+[SYS_link]    "link",
+[SYS_mkdir]   "mkdir",
+[SYS_close]   "close",
+[SYS_trace]   "trace",
+[SYS_history] "history"
+ };
+
+struct syscall_stat{
+  char syscall_name[16];
+  int count;
+  int accum_time;
 };
 
+struct syscall_stat stats[24];
+struct spinlock stats_lock[24];
+
+void init_stats_lock() {
+  for(int i = 1; i < 24; i++){
+    initlock(&stats_lock[i], "history stat");
+    strncpy(stats[i].syscall_name, syscall_names[i], strlen(syscall_names[i]));
+    stats[i].syscall_name[strlen(syscall_names[i])] = '\0';
+  }
+}
+
+int start_tick;
+int end_tick;
+
+void update_history(int syscall_num, int syscall_time, const char* syscall_name){
+  acquire(&stats_lock[syscall_num]);
+
+  if (syscall_num >= 0 && syscall_num < 24) {
+    stats[syscall_num].count++;
+    stats[syscall_num].accum_time += syscall_time;
+  }
+
+  release(&stats_lock[syscall_num]);
+}
+
+int
+get_history(int syscall_num, uint64 addr){
+  struct proc *p = myproc();
+  if(copyout(p->pagetable, addr, (char *)&stats[syscall_num], sizeof(struct syscall_stat)) < 0)
+      return -1;
+  return 0;
+}
+
 void
 syscall(void)
 {
+
   int num;
   struct proc *p = myproc();
 
+  int j;
+  for(j=0;j<6;j++){
+    arg_array[p->pid].exists[j]=false;
+  }
+
   num = p->trapframe->a7;
   if(num > 0 && num < NELEM(syscalls) && syscalls[num]) {
     // Use num to lookup the system call function for num, call it,
     // and store its return value in p->trapframe->a0
-    p->trapframe->a0 = syscalls[num]();
+    acquire(&tickslock);
+    start_tick = ticks;
+    release(&tickslock);
+    int returnValue = syscalls[num]();
+    p->trapframe->a0 = returnValue;
+    acquire(&tickslock);
+    end_tick = ticks;
+    release(&tickslock);
+    update_history(num,(end_tick-start_tick),syscall_names[num]);
+
+       // Print the details of the traced process in the running process
+    if (num == p->traced_pid) {
+      struct syscall_args traced_args = arg_array[p->pid];
+      printf("pid: %d, syscall: %s, ", sys_getpid(), syscall_names[num]);
+      printf("args: (");
+      for (int i = 0; i < 6; i++) {
+         //if(traced_args.args[i].int_arg==0 && traced_args.type[i]!= ARG_INT)
+         //continue;
+         if(traced_args.exists[i]==true){
+          if (i != 5 && i>0)
+          printf(", ");
+        switch (traced_args.type[i]) {
+          case ARG_INT:
+            printf("%d", traced_args.args[i].int_arg);
+          break;
+          case ARG_PTR:
+            printf("%p", traced_args.args[i].ptr_arg);
+            break;
+          case ARG_STR:
+            printf("%s", traced_args.args[i].str_arg);
+            break;
+          default:
+            break;
+        }
+         }
+      }
+      printf("), ");
+      printf("return: %d\n", returnValue);
+
+    }
   } else {
     printf("%d %s: unknown sys call %d\n",
             p->pid, p->name, num);
diff --git a/kernel/syscall.h b/kernel/syscall.h
index bc5f356..f8c6691 100644
--- a/kernel/syscall.h
+++ b/kernel/syscall.h
@@ -1,22 +1,24 @@
 // System call numbers
-#define SYS_fork    1
-#define SYS_exit    2
-#define SYS_wait    3
-#define SYS_pipe    4
-#define SYS_read    5
-#define SYS_kill    6
-#define SYS_exec    7
-#define SYS_fstat   8
-#define SYS_chdir   9
-#define SYS_dup    10
-#define SYS_getpid 11
-#define SYS_sbrk   12
-#define SYS_sleep  13
-#define SYS_uptime 14
-#define SYS_open   15
-#define SYS_write  16
-#define SYS_mknod  17
-#define SYS_unlink 18
-#define SYS_link   19
-#define SYS_mkdir  20
-#define SYS_close  21
+#define SYS_fork     1
+#define SYS_exit     2
+#define SYS_wait     3
+#define SYS_pipe     4
+#define SYS_read     5
+#define SYS_kill     6
+#define SYS_exec     7
+#define SYS_fstat    8
+#define SYS_chdir    9
+#define SYS_dup     10
+#define SYS_getpid  11
+#define SYS_sbrk    12
+#define SYS_sleep   13
+#define SYS_uptime  14
+#define SYS_open    15
+#define SYS_write   16
+#define SYS_mknod   17
+#define SYS_unlink  18
+#define SYS_link    19
+#define SYS_mkdir   20
+#define SYS_close   21
+#define SYS_trace   22
+#define SYS_history 23
\ No newline at end of file
diff --git a/kernel/sysproc.c b/kernel/sysproc.c
index 1de184e..e041f95 100644
--- a/kernel/sysproc.c
+++ b/kernel/sysproc.c
@@ -89,3 +89,20 @@ sys_uptime(void)
   release(&tickslock);
   return xticks;
 }
+
+uint64
+sys_trace(void) {
+  int pid;
+  argint(0, &pid);
+  return get_trace(pid);
+}
+
+uint64
+sys_history(void) {
+  int num;
+  uint64 stat;
+
+  argint(0, &num);
+  argaddr(1,&stat);
+  return get_history(num,stat);
+}
diff --git a/user/history.c b/user/history.c
new file mode 100644
index 0000000..33fccbf
--- /dev/null
+++ b/user/history.c
@@ -0,0 +1,39 @@
+#include "kernel/param.h"
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+struct syscall_stat{
+  char syscall_name[16];
+  int count;
+  int accum_time;
+};
+
+int
+main(int argc, char *argv[]){
+  
+  struct syscall_stat stat;
+
+    if(argc<2){
+        for(int i=1;i<24;i++){
+            if(history(i,&stat)<0){
+                fprintf(2, "%s: history failed\n",argv[0]);
+                exit(0);
+            }
+            else{
+                printf("%d:\tsyscall: %s, #: %d, time: %d\n",i, stat.syscall_name, stat.count, stat.accum_time);
+            }
+        }
+    }
+    else{
+        int syscall_number = atoi(argv[1]);
+        if(history(syscall_number,&stat)<0){
+                fprintf(2, "%s: history failed\n",argv[0]);
+                exit(0);
+            }
+            else{
+                printf("%d:\tsyscall: %s, #: %d, time: %d\n",syscall_number, stat.syscall_name, stat.count, stat.accum_time);
+            }
+    }
+    exit(0);
+}
\ No newline at end of file
diff --git a/user/trace.c b/user/trace.c
new file mode 100755
index 0000000..3dd9b4c
--- /dev/null
+++ b/user/trace.c
@@ -0,0 +1,27 @@
+#include "kernel/param.h"
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int
+main(int argc, char *argv[])
+{
+  int i;
+  char *nargv[MAXARG];
+
+  if(argc < 3 || (argv[1][0] < '0' || argv[1][0] > '9')){
+    fprintf(2, "Usage: %s sys_call_num command\n", argv[0]);
+    exit(1);
+  }
+
+  if (trace(atoi(argv[1])) < 0) {
+    fprintf(2, "%s: trace failed\n", argv[0]);
+    exit(1);
+  }
+
+  for(i = 2; i < argc && i < MAXARG; i++){
+    nargv[i-2] = argv[i];
+  }
+  exec(nargv[0], nargv);
+  exit(0);
+}
diff --git a/user/user.h b/user/user.h
index 4d398d5..bfda961 100644
--- a/user/user.h
+++ b/user/user.h
@@ -1,4 +1,5 @@
 struct stat;
+struct syscall_stat;
 
 // system calls
 int fork(void);
@@ -22,6 +23,8 @@ int getpid(void);
 char* sbrk(int);
 int sleep(int);
 int uptime(void);
+int trace(int);
+int history(int num, struct syscall_stat*);
 
 // ulib.c
 int stat(const char*, struct stat*);
diff --git a/user/usys.pl b/user/usys.pl
index 01e426e..e2095e4 100755
--- a/user/usys.pl
+++ b/user/usys.pl
@@ -36,3 +36,5 @@ entry("getpid");
 entry("sbrk");
 entry("sleep");
 entry("uptime");
+entry("trace");
+entry("history");
